### PR TITLE
Counterexample Improvements

### PR DESCRIPTION
**Pull request by ArquintL**
Supersedes #543 with the same feature additions (support for domain values and functions in counterexamples) but with code formatting improvements.
The unit tests could be improved to check for particular counterexamples as soon as [Silver PR # 516](https://github.com/viperproject/silver/pull/516) is merged.

### CODE REVIEW

**Comment by ArquintL** [comment]
@mschwerhoff I had to create this PR as I cannot push any changes to the existing PR. You might want to look only at the more recent commits as you have already reviewed several commits (look in the old PR)

**Comment by ArquintL** [comment]
CI currently fails as it needs the changes from [Silver PR #551](https://github.com/viperproject/silver/pull/551)

File: src/main/scala/interfaces/Verification.scala
Comment (mschwerhoff) [inline src/main/scala/interfaces/Verification.scala:152]: I find `... map { case (foo, bar) => (foo, interpret(bar))}` to be more readable than `y._1 ... y._2` due to the names given to the two tuple components.

File: src/main/scala/interfaces/Verification.scala
Comment (mschwerhoff) [inline src/main/scala/interfaces/Verification.scala:154]: See previous comment

File: src/main/scala/reporting/DomainInterpreter.scala
Comment (mschwerhoff) [inline src/main/scala/reporting/DomainInterpreter.scala:12]: Please stick with an indentation of 2 (instead of 4). Probably affects several files.

File: src/main/scala/reporting/DomainInterpreter.scala
Comment (mschwerhoff) [inline src/main/scala/reporting/DomainInterpreter.scala:36]: I don't see explicit `x.apply(y)` call very often; just `x(y)` appears to be more common.

File: src/main/scala/reporting/DomainInterpreter.scala
Comment (mschwerhoff) [inline src/main/scala/reporting/DomainInterpreter.scala:38]: apply?

File: src/main/scala/reporting/DomainInterpreter.scala
Comment (mschwerhoff) [inline src/main/scala/reporting/DomainInterpreter.scala:47]: x._1

File: src/test/resources/demo-counterexamples/cyclic-ref.vpr
Comment (mschwerhoff) [inline src/test/resources/demo-counterexamples/cyclic-ref.vpr:1]: Shouldn't there be a copyright message at the beginning of each test file? I guess we're not very consistent here, though.

**Review by mschwerhoff — APPROVED** [review]
Looks generally good, but please address the comments before merging the PR.

**Comment by ArquintL** [comment]
@mschwerhoff thanks for the review! Your comments are all addressed. In addition, I've tried to add some meaningful tests for the provided examples. You can find the corresponding changeset [here](https://github.com/viperproject/silicon/pull/585/files#diff-2eebf6198d69e46e6a6577ca6d3865c785c305f48afcfeea559721b2b6c52d33). What do you think?

**Comment by Aurel300** [comment]
Maybe this should be rebased since the commits by Cedric were already merged in #528? The last couple of commits from that (Cedric's) PR are actually newer that when this (Fabio's) branch started.

**Comment by ArquintL** [comment]
> Maybe this should be rebased since the commits by Cedric were already merged in #528? The last couple of commits from that (Cedric's) PR are actually newer that when this (Fabio's) branch started.

I've already merged in `master` into this PR. I'm not sure why Cedric's commits thus still appear as part of this PR. Do you have an idea how to fix this or should we rather leave it as is and merge this PR?

**Review by mschwerhoff — APPROVED** [review]
Looks good, modulo two copyright headers

**Comment by ArquintL** [comment]
@Aurel300 I've now cleaned up the history and rebased it. I've compared the diff before and after rebasing to make sure everything turned out as expected. Could you maybe take a quick look at the changeset to make sure nothing of Cedric's work is dropped by accident?

File: src/main/scala/reporting/Converter.scala
Comment (Aurel300) [inline src/main/scala/reporting/Converter.scala:195]: Why are these commented out?

File: src/main/scala/reporting/Converter.scala
Comment (Aurel300) [inline src/main/scala/reporting/Converter.scala:226]: Is the comment referring to the `.split("!")` bit? Would be better to make this explicit and searchable:
```suggestion
          // TODO: remove this string-based operation
          case ConstantEntry(v) => DomainValueEntry(id.toString, v.split("!").last)
```

File: src/main/scala/reporting/Converter.scala
Comment (Aurel300) [inline src/main/scala/reporting/Converter.scala:229]: @mschwerhoff or @ArquintL does the comment here make sense? When can this happen?

File: src/main/scala/reporting/Converter.scala
Comment (Aurel300) [inline src/main/scala/reporting/Converter.scala:321]: Is this a typo? Should it be `($to)` instead?

File: src/main/scala/reporting/Converter.scala
Comment (Aurel300) [inline src/main/scala/reporting/Converter.scala:354]: Is this hardcoded to only look for 10 references?

File: src/main/scala/reporting/Converter.scala
Comment (Aurel300) [inline src/main/scala/reporting/Converter.scala:406]: This looks like a very simple, stripped-down interpreter for counterexample terms. Is there anything in Silicon that could already do this? If not, could this be useful for more than just evaluation permissions?

File: src/main/scala/reporting/Converter.scala
Comment (Aurel300) [inline src/main/scala/reporting/Converter.scala:422]: Maybe something like this for all of the operations?

```suggestion
      case PermTimes(v1, v2) =>
        evalPerm(v1, model).map(x => evalPerm(v2, model).map(y => x * y))
```

File: src/main/scala/reporting/Converter.scala
Comment (Aurel300) [inline src/main/scala/reporting/Converter.scala:578]: Are we sure this isn't hiding errors we would like to see?

**Comment by Aurel300** [comment]
Do not merge this: @mlimbeck will open another PR based on this one with some fixes.

**Pull request merged**
